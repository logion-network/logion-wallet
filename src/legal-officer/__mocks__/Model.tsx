import {
    fetchRecoveryInfo,
    acceptAccountRecoveryRequest,
    rejectAccountRecoveryRequest,
} from './ModelMock';

export {
    fetchRecoveryInfo,
    acceptAccountRecoveryRequest,
    rejectAccountRecoveryRequest,
};

export function toRecoveryRequestType(value: string | undefined) {
    return value;
}
