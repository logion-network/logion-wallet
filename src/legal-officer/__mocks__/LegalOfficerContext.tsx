import {
    useLegalOfficerContext,
} from "./LegalOfficerContextMock";

export {
    useLegalOfficerContext
}
