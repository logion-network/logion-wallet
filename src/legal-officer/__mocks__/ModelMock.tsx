export let fetchRecoveryInfo = jest.fn();

export function setFetchRecoveryInfo(mockFn: any) {
    fetchRecoveryInfo = mockFn;
}

export let rejectAccountRecoveryRequest = jest.fn();

export function setRejectAccountRecoveryRequest(mockFn: any) {
    rejectAccountRecoveryRequest = mockFn;
}

export let acceptAccountRecoveryRequest = jest.fn();

export function setAcceptAccountRecoveryRequest(mockFn: any) {
    acceptAccountRecoveryRequest = mockFn;
}
