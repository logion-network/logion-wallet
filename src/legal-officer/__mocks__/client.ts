import {
    closeLoc,
    deleteLink,
    voidLoc,
    getVerifiedThirdPartySelections,
} from "./ClientMock";

export {
    closeLoc,
    deleteLink,
    voidLoc,
    getVerifiedThirdPartySelections,
}
