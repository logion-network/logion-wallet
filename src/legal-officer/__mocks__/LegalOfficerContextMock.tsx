import { LocType, IdentityLocType } from "@logion/node-api/dist/Types";
import { AxiosInstance } from "axios";

import { COLOR_THEME } from "src/common/TestData";
import { RequestAndLoc } from "../LegalOfficerContext";
import { LegalOfficerData } from "../LegalOfficerData";

export let pendingTokenizationRequests: any[] | null = null;

export let rejectRequest = () => {

};

export let tokenizationRequestsHistory: any[] | null = null;

export let refreshRequests = jest.fn();

export let pendingProtectionRequests: any[] | null = null;

export let activatedProtectionRequests: any[] | null = null;

export let protectionRequestsHistory: any[] | null = null;

export let pendingRecoveryRequests: any[] | null = null;

export let recoveryRequestsHistory: any[] | null = null;

export function setPendingRequests(requests: any[]) {
    pendingTokenizationRequests = requests;
}

export function setRejectRequest(callback: any) {
    rejectRequest = callback;
}

export function setTokenizationRequestsHistory(requests: any[]) {
    tokenizationRequestsHistory = requests;
}

export function setRefreshRequests(callback: any) {
    refreshRequests = callback;
}

export function setPendingProtectionRequests(requests: any[]) {
    pendingProtectionRequests = requests;
}

export function setActivatedProtectionRequests(requests: any[]) {
    activatedProtectionRequests = requests;
}

export function setProtectionRequestsHistory(requests: any[]) {
    protectionRequestsHistory = requests;
}

export function setPendingRecoveryRequests(requests: any[]) {
    pendingRecoveryRequests = requests;
}

export function setRecoveryRequestsHistory(requests: any[]) {
    recoveryRequestsHistory = requests;
}

export let pendingLocRequests: Record<LocType, any[]> | null = null;
export let rejectedLocRequests: Record<LocType, any[]> | null = null;

export let openedLocRequests: Record<LocType, any[]> | null = {
    "Transaction": [],
    "Collection": [],
    "Identity": [],
};

export let closedLocRequests: Record<LocType, any[]> | null = {
    "Transaction": [],
    "Collection": [],
    "Identity": [],
};

export let openedIdentityLocs: RequestAndLoc[] | null = null;

export let openedIdentityLocsByType: Record<IdentityLocType, RequestAndLoc[]> | null = null;

export let closedIdentityLocsByType: Record<IdentityLocType, RequestAndLoc[]> | null = null;

export let voidIdentityLocsByType: Record<IdentityLocType, RequestAndLoc[]> | null = null;

export function setRejectedLocRequests(requests: any[]) {
    rejectedLocRequests = { Collection: [], Transaction: requests, Identity: [] };
}

export function setOpenedLocRequests(requests: any[]) {
    openedLocRequests = { Collection: [], Transaction: requests, Identity: [] };
}

export function setPendingLocRequests(requests: any[]) {
    pendingLocRequests = { Collection: [], Transaction: requests, Identity: [] };
}

export function setClosedLocRequests(requests: any[]) {
    closedLocRequests = { Collection: [], Transaction: requests, Identity: [] };
}

export function setOpenedIdentityLocs(requests: any[]) {
    openedIdentityLocs = requests;
}

export function setOpenedIdentityLocsByType(locs: Record<IdentityLocType, RequestAndLoc[]>) {
    openedIdentityLocsByType = locs;
}

export function setClosedIdentityLocsByType(locs: Record<IdentityLocType, RequestAndLoc[]>) {
    closedIdentityLocsByType = locs;
}

export function setVoidedIdentityLocsByType(locs: Record<IdentityLocType, RequestAndLoc[]>) {
    voidIdentityLocsByType = locs;
}

let onchainSettings: LegalOfficerData = {};

export function setOnchainSettings(settings: LegalOfficerData) {
    onchainSettings = settings;
}

export let refreshLocs = jest.fn();

export function useLegalOfficerContext() {
    return {
        pendingTokenizationRequests,
        rejectRequest,
        tokenizationRequestsHistory,
        refreshRequests,
        pendingProtectionRequests,
        activatedProtectionRequests,
        protectionRequestsHistory,
        colorTheme: COLOR_THEME,
        openedLocRequests,
        openedIdentityLocs,
        openedIdentityLocsByType,
        closedLocRequests,
        closedIdentityLocsByType,
        pendingLocRequests,
        rejectedLocRequests,
        voidIdentityLocsByType,
        axios: {} as AxiosInstance,
        onchainSettings,
        pendingRecoveryRequests,
        refreshLocs,
    };
}
