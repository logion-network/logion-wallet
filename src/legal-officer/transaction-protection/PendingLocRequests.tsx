import { LocType } from "@logion/node-api";

import Table, { Cell, EmptyTableMessage, DateTimeCell } from '../../common/Table';
import LocStatusCell from '../../common/LocStatusCell';
import Button from '../../common/Button';

import ButtonGroup from "../../common/ButtonGroup";
import UserIdentityNameCell from '../../common/UserIdentityNameCell';

import LocRequestDetails from './LocRequestDetails';
import { useLegalOfficerContext } from "../LegalOfficerContext";
import { useResponsiveContext } from "../../common/Responsive";
import { locDetailsPath } from "../LegalOfficerPaths";
import { useNavigate } from 'react-router-dom';
import { useMemo } from "react";

export interface Props {
    locType: LocType;
}

export default function PendingLocRequests(props: Props) {
    const { pendingLocRequests } = useLegalOfficerContext();
    const { locType } = props;
    const { width } = useResponsiveContext();
    const navigate = useNavigate();

    const data = useMemo(() => pendingLocRequests ? pendingLocRequests[locType].map(loc => loc.data()) : [], [ pendingLocRequests, locType ]);

    if (pendingLocRequests === null) {
        return null;
    }

    return (
        <>
            <Table
                columns={[
                    {
                        header: "Requester",
                        render: request => <UserIdentityNameCell userIdentity={ request.userIdentity } />,
                        align: "left",
                        renderDetails: request => <LocRequestDetails request={ request }/>
                    },
                    {
                        header: "Description",
                        render: request => <Cell content={ request.description } overflowing tooltipId="description" />,
                        align: "left",
                    },
                    {
                        header: "Status",
                        render: request => <LocStatusCell status={ request.status }/>,
                        width: width({
                            onSmallScreen: locType === "Identity" ? "140px" : "130px",
                            otherwise: "140px",
                        }),
                    },
                    {
                        header: "Creation Date",
                        render: request => <DateTimeCell dateTime={ request.createdOn || null } />,
                        width: width({
                            onSmallScreen: locType === "Identity" ? "200px" : "130",
                            otherwise: "200px",
                        }),
                        align: 'center',
                    },
                    {
                        header: "",
                        render: request => (
                            <ButtonGroup aria-label="actions">
                                <Button
                                    onClick={ () => navigate(locDetailsPath(request.id, request.locType)) }
                                >
                                    Review
                                </Button>
                            </ButtonGroup>
                        ),
                        width: "150px",
                    }
                ]}
                data={ data }
                renderEmpty={ () => <EmptyTableMessage>No pending LOC request</EmptyTableMessage> }
            />
        </>
    );
}
