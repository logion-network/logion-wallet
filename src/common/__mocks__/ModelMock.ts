import { LocRequest } from "@logion/client";
import { CreateLocRequest } from "../Model";

export let fetchProtectionRequests = jest.fn();

export function setFetchProtectionRequests(mockFn: any) {
    fetchProtectionRequests = mockFn;
}

export async function createLocRequest(axios: any, request: CreateLocRequest): Promise<LocRequest> {
    return Promise.resolve({
        ...request,
        createdOn: "2022-01-20T09:24:00.000",
        description: "Test",
        files: [],
        id: "671e65cf-42c3-435b-a222-be008e7fb3d1",
        links: [],
        metadata: [],
        status: "OPEN",
        verifiedThirdParty: false,
        selectedIssuers: [],
    });
}

export const preVoid = jest.fn();

export function resetDefaultMocks() {
    preVoid.mockResolvedValue(undefined);
}

export function isGrantedAccess() {
    return true;
}
