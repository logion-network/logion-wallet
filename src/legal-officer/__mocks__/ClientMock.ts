import { OpenLoc, VerifiedThirdParty } from "@logion/client";

export let closeLoc = jest.fn().mockResolvedValue(undefined);

export function setCloseLocMock(mock: any) {
    closeLoc = mock;
}

export let deleteLink = jest.fn().mockResolvedValue(undefined);

export let voidLoc = jest.fn().mockResolvedValue(undefined);

export function setVoidLocMock(mock: any) {
    voidLoc = mock;
}

let vtpSelections: VerifiedThirdParty[] = [];

export async function getVerifiedThirdPartySelections(params: { locState: OpenLoc } ): Promise<VerifiedThirdParty[]> {
    return Promise.resolve(vtpSelections);
}

export function setVerifiedThirdPartySelections(mock: VerifiedThirdParty[]) {
    vtpSelections = mock;
}

export let requestVote = jest.fn().mockResolvedValue(undefined);

export function setRequestVoteMock(mock: any) {
    requestVote = mock;
}
