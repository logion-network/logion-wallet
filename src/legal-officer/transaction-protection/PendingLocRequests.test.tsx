jest.mock('../../logion-chain');
jest.mock('../../common/CommonContext');
jest.mock('../LegalOfficerContext');
jest.mock('../../loc/Model');

import { render, screen, waitFor, getByText } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { shallowRender } from '../../tests';

import { axiosMock, setCurrentAddress, DEFAULT_LEGAL_OFFICER_ACCOUNT } from '../../logion-chain/__mocks__/LogionChainMock';
import { setPendingLocRequests } from "../__mocks__/LegalOfficerContextMock";

import { setRejectLocRequest } from '../../loc/__mocks__/ModelMock';

import PendingLocRequests from './PendingLocRequests';

beforeEach(() => {
    setCurrentAddress(DEFAULT_LEGAL_OFFICER_ACCOUNT);
});

describe("PendingLocRequests", () => {

    it("Renders null with no data", () => {
        const tree = shallowRender(<PendingLocRequests locType="Transaction" />);
        expect(tree).toMatchSnapshot();
    });

    it("Renders pending requests", () => {
        setPendingLocRequests([
            {
                id: "1",
                ownerAddress: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
                requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
                status: "REQUESTED"
            }
        ]);
        const tree = shallowRender(<PendingLocRequests locType="Transaction" />);
        expect(tree).toMatchSnapshot();
});

    it("Click on reject and confirm rejects request", async () => {
        setPendingLocRequests([
            {
                id: "1",
                ownerAddress: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
                requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
                status: "REQUESTED"
            }
        ]);
        const rejectCallback = jest.fn();
        setRejectLocRequest(rejectCallback);

    render(<PendingLocRequests locType="Transaction" />);
    const rejectButton = screen.getByTestId("reject-1");
    await userEvent.click(rejectButton);

    const reasonText = "Because";
    const reasonTextArea = screen.getByTestId("reason");
    await userEvent.type(reasonTextArea, reasonText);

    const confirmButton = screen.getByRole("button", {name: "Proceed"});
    await userEvent.click(confirmButton);

    expect(rejectCallback).toBeCalledWith(
        axiosMock.object(),
        expect.objectContaining({
            requestId: "1",
            rejectReason: reasonText

        })
    );
    await waitFor(() => expect(screen.queryByTestId("modal-reject-1")).not.toBeInTheDocument());
});

    it("Click on accept opens acceptance process", async () => {
        setPendingLocRequests([
            {
                id: "1",
                legalOfficerAddress: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
                requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
                requestedTokenName: "TOKEN1",
                bars: 1,
                status: "PENDING",
                locType: "Transaction"
            }
    ]);

        const tree = render(<PendingLocRequests locType="Transaction" />);

        const acceptButton = tree.getByTestId("accept-1");
        await userEvent.click(acceptButton);

        const acceptingModal = tree.getByRole('dialog');
        expect(getByText(acceptingModal, "Accepting Transaction Protection Request")).toBeInTheDocument();
        expect(acceptingModal).toBeInTheDocument();
    });
})
