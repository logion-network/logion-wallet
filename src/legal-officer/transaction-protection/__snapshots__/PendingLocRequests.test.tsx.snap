// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PendingLocRequests Renders null with no data 1`] = `null`;

exports[`PendingLocRequests Renders pending requests 1`] = `
<React.Fragment>
  <Table
    columns={
      Array [
        Object {
          "align": "left",
          "header": "Requester",
          "render": [Function],
          "renderDetails": [Function],
        },
        Object {
          "align": "left",
          "header": "Description",
          "render": [Function],
        },
        Object {
          "header": "Status",
          "render": [Function],
          "width": "140px",
        },
        Object {
          "align": "center",
          "header": "Creation Date",
          "render": [Function],
          "width": "200px",
        },
        Object {
          "header": "",
          "render": [Function],
          "width": "320px",
        },
      ]
    }
    data={
      Array [
        Object {
          "id": "1",
          "ownerAddress": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "requesterAddress": "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
          "status": "REQUESTED",
        },
      ]
    }
    renderEmpty={[Function]}
  />
  <LocRequestAcceptance
    clearRequestToAccept={[Function]}
    requestToAccept={null}
  />
</React.Fragment>
`;
